tidy3d/components/data/monitor_data.py

```diff
@@ -424,6 +424,194 @@ class ModeSolverData(ModeSolverDataset, ElectromagneticFieldData):
 
     monitor: ModeSolverMonitor
 
+    # pylint:disable=too-many-locals
+    def overlap_sort(
+        self,
+        track: Literal["first", "last", "central"],
+        overlap_thresh: float = 0.75,
+    ) -> ModeSolverData:
+        """Starting from the base frequency defined by parameter ``track``, sort modes at each
+        frequency according to their overlap values with the modes at the previous frequency.
+        That is, it attempts to rearrange modes in such a way that a given ``mode_index``
+        corresponds to physically the same mode at all frequencies.
+
+        Parameters
+        ----------
+        track : Literal['first', 'last', 'central']
+            Parameter that specifies which frequency will serve as a starting point in
+            the reordering process.
+        overlap_thresh : float = 0.75
+            Modal overlap threshold above which two modes are considered to be the same. It serves
+            two purposes. First, if overlap values for corresponding modes of two neighboring
+            frequencies are all above the provided value then no reordering is performed between
+            the two frequencies. Second, if after reordering the modal overlap between two
+            frequencies of the same ``mode_index`` is still less than the provided value then
+            a warning about a possible discontinuity is raised.
+        """
+        num_freqs = len(self.monitor.freqs)
+        num_modes = self.monitor.mode_spec.num_modes
+
+        if track == "first":
+            f0_ind = 0
+        elif track == "last":
+            f0_ind = num_freqs - 1
+        elif track == "center":
+            f0_ind = num_freqs // 2
+
+        # For each mode index determine location for gauge selection
+        gauge_indices = np.zeros(num_modes, dtype=int)
+        for mode_idx in range(num_modes):
+            e_field_all = np.concatenate(
+                (
+                    self.Ex.isel(f=f0_ind, mode_index=mode_idx).data.ravel(),
+                    self.Ey.isel(f=f0_ind, mode_index=mode_idx).data.ravel(),
+                    self.Ez.isel(f=f0_ind, mode_index=mode_idx).data.ravel(),
+                )
+            )
+            gauge_indices[mode_idx] = np.argmax(np.abs(e_field_all))
+
+        # Compute sorting order, overlap with neighbor
+        sorting = -np.ones((num_freqs, num_modes), dtype=int)
+        overlap = np.zeros((num_freqs, num_modes))
+        sorting[f0_ind, :] = np.arange(num_modes)
+
+        # Sort in two directions from the base frequency
+        for step, last_ind in zip([-1, 1], [-1, num_freqs]):
+            # Start with the base frequency (storing norm to avoid double work)
+            data_template = self.sel_freq_index(freq_index=f0_ind)
+            norm_template = np.sqrt(np.abs(data_template.dot(data_template)))
```

Comment:
Let's add the normalization directly in the mode solver so that all the `norm` stuff here can be removed. That is to say let's make the mode solver return normalized modes.